#### Data: Dry Bean

In [1]:
import pandas as pd
import numpy as np

from ucimlrepo import fetch_ucirepo 
  
# fetch dataset 
dry_bean = fetch_ucirepo(id=602) 
  
# data (as pandas dataframes) 
X = dry_bean.data.features 
y = dry_bean.data.targets 

In [2]:
X.shape, y.shape

((13611, 16), (13611, 1))

In [3]:
# concating the two data extracted from the source
data = pd.concat([X,y], axis = 1)

In [4]:
data.shape

(13611, 17)

In [5]:
data.head(3)

Area  Perimeter  MajorAxisLength  MinorAxisLength  AspectRatio  \
0  28395    610.291       208.178117       173.888747     1.197191   
1  28734    638.018       200.524796       182.734419     1.097356   
2  29380    624.110       212.826130       175.931143     1.209713   

   Eccentricity  ConvexArea  EquivDiameter    Extent  Solidity  Roundness  \
0      0.549812       28715     190.141097  0.763923  0.988856   0.958027   
1      0.411785       29172     191.272751  0.783968  0.984986   0.887034   
2      0.562727       29690     193.410904  0.778113  0.989559   0.947849   

   Compactness  ShapeFactor1  ShapeFactor2  ShapeFactor3  ShapeFactor4  Class  
0     0.913358      0.007332      0.003147      0.834222      0.998724  SEKER  
1     0.953861      0.006979      0.003564      0.909851      0.998430  SEKER  
2     0.908774      0.007244      0.003048      0.825871      0.999066  SEKER

In [6]:
data.columns

Index(['Area', 'Perimeter', 'MajorAxisLength', 'MinorAxisLength',
       'AspectRatio', 'Eccentricity', 'ConvexArea', 'EquivDiameter', 'Extent',
       'Solidity', 'Roundness', 'Compactness', 'ShapeFactor1', 'ShapeFactor2',
       'ShapeFactor3', 'ShapeFactor4', 'Class'],
      dtype='object')

In [7]:
data.isnull().sum()

Area               0
Perimeter          0
MajorAxisLength    0
MinorAxisLength    0
AspectRatio        0
Eccentricity       0
ConvexArea         0
EquivDiameter      0
Extent             0
Solidity           0
Roundness          0
Compactness        0
ShapeFactor1       0
ShapeFactor2       0
ShapeFactor3       0
ShapeFactor4       0
Class              0
dtype: int64

In [8]:
data.Class.value_counts()

Class
DERMASON    3546
SIRA        2636
SEKER       2027
HOROZ       1928
CALI        1630
BARBUNYA    1322
BOMBAY       522
Name: count, dtype: int64

In [9]:
unseen_data = data.sample(frac=0.15, random_state=42)
df = data.drop(unseen_data.index)

In [10]:
unseen_data.to_csv('./Project/drybean_unseen.csv', index=False)
df.to_csv('./Project/Dry_Bean.csv', index=False)

In [11]:
data = pd.read_csv('./Project/Dry_Bean.csv')

In [12]:
data.head(1)

Area  Perimeter  MajorAxisLength  MinorAxisLength  AspectRatio  \
0  28395    610.291       208.178117       173.888747     1.197191   

   Eccentricity  ConvexArea  EquivDiameter    Extent  Solidity  Roundness  \
0      0.549812       28715     190.141097  0.763923  0.988856   0.958027   

   Compactness  ShapeFactor1  ShapeFactor2  ShapeFactor3  ShapeFactor4  Class  
0     0.913358      0.007332      0.003147      0.834222      0.998724  SEKER

In [13]:
from pycaret.classification import ClassificationExperiment
s = ClassificationExperiment()
s.setup(data, target = 'Class', session_id = 42)

In [14]:
best = s.compare_models()

In [15]:
s.models()

Name  \
ID                                          
lr                    Logistic Regression   
knn                K Neighbors Classifier   
nb                            Naive Bayes   
dt               Decision Tree Classifier   
svm                   SVM - Linear Kernel   
rbfsvm                SVM - Radial Kernel   
gpc           Gaussian Process Classifier   
mlp                        MLP Classifier   
ridge                    Ridge Classifier   
rf               Random Forest Classifier   
qda       Quadratic Discriminant Analysis   
ada                  Ada Boost Classifier   
gbc          Gradient Boosting Classifier   
lda          Linear Discriminant Analysis   
et                 Extra Trees Classifier   
lightgbm  Light Gradient Boosting Machine   
catboost              CatBoost Classifier   
dummy                    Dummy Classifier   

                                                  Reference  Turbo  
ID                                                                  
lr        sklearn.linear_model._logistic.LogisticRegression   True  
knn       sklearn.neighbors._classification.KNeighborsCl...   True  
nb                           sklearn.naive_bayes.GaussianNB   True  
dt             sklearn.tree._classes.DecisionTreeClassifier   True  
svm       sklearn.linear_model._stochastic_gradient.SGDC...   True  
rbfsvm                             sklearn.svm._classes.SVC  False  
gpc       sklearn.gaussian_process._gpc.GaussianProcessC...  False  
mlp       sklearn.neural_network._multilayer_perceptron....  False  
ridge           sklearn.linear_model._ridge.RidgeClassifier   True  
rf          sklearn.ensemble._forest.RandomForestClassifier   True  
qda       sklearn.discriminant_analysis.QuadraticDiscrim...   True  
ada       sklearn.ensemble._weight_boosting.AdaBoostClas...   True  
gbc         sklearn.ensemble._gb.GradientBoostingClassifier   True  
lda       sklearn.discriminant_analysis.LinearDiscrimina...   True  
et            sklearn.ensemble._forest.ExtraTreesClassifier   True  
lightgbm                    lightgbm.sklearn.LGBMClassifier   True  
catboost                   catboost.core.CatBoostClassifier   True  
dummy                         sklearn.dummy.DummyClassifier   True

In [16]:
lightgbm = s.create_model('lightgbm')

In [17]:
lightgbm

LGBMClassifier(boosting_type='gbdt', class_weight=None, colsample_bytree=1.0,
               importance_type='split', learning_rate=0.1, max_depth=-1,
               min_child_samples=20, min_child_weight=0.001, min_split_gain=0.0,
               n_estimators=100, n_jobs=-1, num_leaves=31, objective=None,
               random_state=42, reg_alpha=0.0, reg_lambda=0.0, subsample=1.0,
               subsample_for_bin=200000, subsample_freq=0)

In [18]:
tune_lightgbm = s.tune_model(lightgbm)

Fitting 10 folds for each of 10 candidates, totalling 100 fits


In [19]:
final_lightgbm = s.finalize_model(tune_lightgbm)

In [20]:
final_lightgbm

Pipeline(memory=Memory(location=None),
         steps=[('label_encoding',
                 TransformerWrapperWithInverse(exclude=None, include=None,
                                               transformer=LabelEncoder())),
                ('numerical_imputer',
                 TransformerWrapper(exclude=None,
                                    include=['Area', 'Perimeter',
                                             'MajorAxisLength',
                                             'MinorAxisLength', 'AspectRatio',
                                             'Eccentricity', 'ConvexArea',
                                             'EquivDiameter', 'Extent',
                                             'Solidity', 'Roundn...
                                boosting_type='gbdt', class_weight=None,
                                colsample_bytree=1.0, feature_fraction=1.0,
                                importance_type='split', learning_rate=0.1,
                                max_depth=-1, min_child_samples=71,
                                min_child_weight=0.001, min_split_gain=0.6,
                                n_estimators=130, n_jobs=-1, num_leaves=4,
                                objective=None, random_state=42, reg_alpha=0.3,
                                reg_lambda=4, subsample=1.0,
                                subsample_for_bin=200000, subsample_freq=0))],
         verbose=False)

In [21]:
s.evaluate_model(final_lightgbm)

interactive(children=(ToggleButtons(description='Plot Type:', icons=('',), options=(('Pipeline Plot', 'pipelin…

In [22]:
data = s.predict_model(final_lightgbm)

[LightGBM] [Warning] feature_fraction is set=1.0, colsample_bytree=1.0 will be ignored. Current value: feature_fraction=1.0
[LightGBM] [Warning] bagging_fraction is set=0.6, subsample=1.0 will be ignored. Current value: bagging_fraction=0.6
[LightGBM] [Warning] bagging_freq is set=5, subsample_freq=0 will be ignored. Current value: bagging_freq=5
[LightGBM] [Warning] feature_fraction is set=1.0, colsample_bytree=1.0 will be ignored. Current value: feature_fraction=1.0
[LightGBM] [Warning] bagging_fraction is set=0.6, subsample=1.0 will be ignored. Current value: bagging_fraction=0.6
[LightGBM] [Warning] bagging_freq is set=5, subsample_freq=0 will be ignored. Current value: bagging_freq=5


In [23]:
data[data['Class'] !=data['prediction_label']]

Area    Perimeter  MajorAxisLength  MinorAxisLength  AspectRatio  \
11080  37307   728.892029       267.065674       179.108551     1.491083   
6304   34189   687.252014       249.062119       174.993652     1.423264   
3280   45666   789.770020       274.863342       211.885132     1.297228   
6139   63083   993.172974       384.437622       210.901138     1.822833   
11243  38471   726.153992       266.091248       184.475708     1.442419   
...      ...          ...              ...              ...          ...   
5800   57371   905.573975       352.953033       209.091858     1.688029   
4766   41486   753.172974       278.307251       190.358139     1.462019   
11113  37514   725.846985       269.881165       177.418228     1.521158   
6285   73451  1081.600952       421.208191       222.836227     1.890214   
31     32026   654.237000       230.979065       176.931808     1.305469   

       Eccentricity  ConvexArea  EquivDiameter    Extent  Solidity  Roundness  \
11080      0.741771       37886     217.946655  0.741631  0.984717   0.882417   
6304       0.711575       34559     208.640335  0.730066  0.989294   0.909629   
3280       0.636987       46074     241.130173  0.748378  0.991145   0.920029   
6139       0.836087       63902     283.407440  0.752996  0.987184   0.803660   
11243      0.720668       38875     221.320572  0.806791  0.989608   0.916824   
...             ...         ...            ...       ...       ...        ...   
5800       0.805639       58240     270.272125  0.674921  0.985079   0.879133   
4766       0.729496       41862     229.829529  0.805366  0.991018   0.919014   
11113      0.753547       37920     218.550476  0.793309  0.989293   0.894773   
6285       0.848597       74535     305.811584  0.644714  0.985457   0.788993   
31         0.642831       32349     201.932587  0.779449  0.990015   0.940248   

       Compactness  ShapeFactor1  ShapeFactor2  ShapeFactor3  ShapeFactor4  \
11080     0.816079      0.007159      0.001959      0.665985      0.993038   
6304      0.837704      0.007285      0.002213      0.701748      0.998772   
3280      0.877273      0.006019      0.002199      0.769608      0.998357   
6139      0.737200      0.006094      0.001110      0.543464      0.990644   
11243     0.831747      0.006917      0.002042      0.691803      0.997870   
...            ...           ...           ...           ...           ...   
5800      0.765745      0.006152      0.001305      0.586366      0.989802   
4766      0.825812      0.006708      0.001925      0.681966      0.997047   
11113     0.809803      0.007194      0.001908      0.655780      0.997545   
6285      0.726034      0.005735      0.000983      0.527126      0.996380   
31        0.874246      0.007212      0.002599      0.764307      0.997779   

          Class prediction_label  prediction_score  
11080  DERMASON             SIRA            0.5022  
6304       SIRA         DERMASON            0.9604  
3280       CALI            SEKER            0.9971  
6139      HOROZ             CALI            0.5806  
11243  DERMASON             SIRA            0.5391  
...         ...              ...               ...  
5800      HOROZ             SIRA            0.3459  
4766      HOROZ             SIRA            0.5735  
11113  DERMASON             SIRA            0.5074  
6285      HOROZ         BARBUNYA            0.6689  
31        SEKER         DERMASON            0.9164  

[215 rows x 19 columns]

In [24]:
data = pd.read_csv('./Project/Dry_Bean.csv')

In [25]:
data.head(1)

Area  Perimeter  MajorAxisLength  MinorAxisLength  AspectRatio  \
0  28395    610.291       208.178117       173.888747     1.197191   

   Eccentricity  ConvexArea  EquivDiameter    Extent  Solidity  Roundness  \
0      0.549812       28715     190.141097  0.763923  0.988856   0.958027   

   Compactness  ShapeFactor1  ShapeFactor2  ShapeFactor3  ShapeFactor4  Class  
0     0.913358      0.007332      0.003147      0.834222      0.998724  SEKER

In [26]:
from pycaret.classification import setup, models, create_model, tune_model, finalize_model, save_model, evaluate_model
cls = setup(data, target='Class', session_id = 42)

In [27]:
lightgbm = create_model('lightgbm')
lightgbm

LGBMClassifier(boosting_type='gbdt', class_weight=None, colsample_bytree=1.0,
               importance_type='split', learning_rate=0.1, max_depth=-1,
               min_child_samples=20, min_child_weight=0.001, min_split_gain=0.0,
               n_estimators=100, n_jobs=-1, num_leaves=31, objective=None,
               random_state=42, reg_alpha=0.0, reg_lambda=0.0, subsample=1.0,
               subsample_for_bin=200000, subsample_freq=0)

In [28]:
tuned_lightgbm = tune_model(lightgbm)

Fitting 10 folds for each of 10 candidates, totalling 100 fits


In [29]:
final_lightgbm = finalize_model(tuned_lightgbm)
final_lightgbm

Pipeline(memory=Memory(location=None),
         steps=[('label_encoding',
                 TransformerWrapperWithInverse(exclude=None, include=None,
                                               transformer=LabelEncoder())),
                ('numerical_imputer',
                 TransformerWrapper(exclude=None,
                                    include=['Area', 'Perimeter',
                                             'MajorAxisLength',
                                             'MinorAxisLength', 'AspectRatio',
                                             'Eccentricity', 'ConvexArea',
                                             'EquivDiameter', 'Extent',
                                             'Solidity', 'Roundn...
                                boosting_type='gbdt', class_weight=None,
                                colsample_bytree=1.0, feature_fraction=1.0,
                                importance_type='split', learning_rate=0.1,
                                max_depth=-1, min_child_samples=71,
                                min_child_weight=0.001, min_split_gain=0.6,
                                n_estimators=130, n_jobs=-1, num_leaves=4,
                                objective=None, random_state=42, reg_alpha=0.3,
                                reg_lambda=4, subsample=1.0,
                                subsample_for_bin=200000, subsample_freq=0))],
         verbose=False)

In [30]:
save_model(final_lightgbm, './Project/Dry_Bean_Unseen')

Transformation Pipeline and Model Successfully Saved


(Pipeline(memory=Memory(location=None),
          steps=[('label_encoding',
                  TransformerWrapperWithInverse(exclude=None, include=None,
                                                transformer=LabelEncoder())),
                 ('numerical_imputer',
                  TransformerWrapper(exclude=None,
                                     include=['Area', 'Perimeter',
                                              'MajorAxisLength',
                                              'MinorAxisLength', 'AspectRatio',
                                              'Eccentricity', 'ConvexArea',
                                              'EquivDiameter', 'Extent',
                                              'Solidity', 'Roundn...
                                 boosting_type='gbdt', class_weight=None,
                                 colsample_bytree=1.0, feature_fraction=1.0,
                                 importance_type='split', learning_rate=0.1,
                      

In [31]:
import joblib

from sklearn.metrics import classification_report
test = pd.read_csv('./Project/drybean_unseen.csv')

In [32]:
test.shape

(2042, 17)

In [33]:
cls = joblib.load('./Project/Dry_Bean_Unseen.pkl')
cls

Pipeline(memory=Memory(location=None),
         steps=[('label_encoding',
                 TransformerWrapperWithInverse(exclude=None, include=None,
                                               transformer=LabelEncoder())),
                ('numerical_imputer',
                 TransformerWrapper(exclude=None,
                                    include=['Area', 'Perimeter',
                                             'MajorAxisLength',
                                             'MinorAxisLength', 'AspectRatio',
                                             'Eccentricity', 'ConvexArea',
                                             'EquivDiameter', 'Extent',
                                             'Solidity', 'Roundn...
                                boosting_type='gbdt', class_weight=None,
                                colsample_bytree=1.0, feature_fraction=1.0,
                                importance_type='split', learning_rate=0.1,
                                max_depth=-1, min_child_samples=71,
                                min_child_weight=0.001, min_split_gain=0.6,
                                n_estimators=130, n_jobs=-1, num_leaves=4,
                                objective=None, random_state=42, reg_alpha=0.3,
                                reg_lambda=4, subsample=1.0,
                                subsample_for_bin=200000, subsample_freq=0))],
         verbose=False)

In [34]:
unseen_X = test.drop('Class', axis=1)
y = test['Class']

In [35]:
y_pred = cls.predict(unseen_X)
y_pred

[LightGBM] [Warning] feature_fraction is set=1.0, colsample_bytree=1.0 will be ignored. Current value: feature_fraction=1.0
[LightGBM] [Warning] bagging_fraction is set=0.6, subsample=1.0 will be ignored. Current value: bagging_fraction=0.6
[LightGBM] [Warning] bagging_freq is set=5, subsample_freq=0 will be ignored. Current value: bagging_freq=5


0          SEKER
1       BARBUNYA
2          SEKER
3          SEKER
4       DERMASON
          ...   
2037    BARBUNYA
2038       HOROZ
2039        SIRA
2040       SEKER
2041       SEKER
Name: Class, Length: 2042, dtype: object

In [36]:
print(classification_report(y, y_pred))

              precision    recall  f1-score   support

    BARBUNYA       0.92      0.91      0.91       204
      BOMBAY       1.00      1.00      1.00        82
        CALI       0.95      0.95      0.95       243
    DERMASON       0.89      0.94      0.92       494
       HOROZ       0.98      0.97      0.98       303
       SEKER       0.97      0.94      0.95       306
        SIRA       0.89      0.86      0.87       410

    accuracy                           0.93      2042
   macro avg       0.94      0.94      0.94      2042
weighted avg       0.93      0.93      0.93      2042



This classification result is showing the performance of a model in predicting different classes in Dry Bean dataset. Here's the explanationa of each metric:

Precision: It measures how many of the predicted instances of a class are actually correct. i.e, for class BARBUNYA, the precision is 0.92, which means that out of all instances predicted as BARBUNYA, 92% were correct.

Recall: It measures how many of the actual instances of a class were predicted correctly by the model. therefore, for class DERMASON, the recall is 0.94, which means that out of all actual instances of DERMASON, 94% were predicted correctly by the model.

F1-score: It is the harmonic mean of precision and recall, providing a single score that balances both metrics. It's useful when we want to consider both precision and recall together.

Support: It indicates the number of actual instances of each class in Dry Bean dataset.

Accuracy: It measures the overall correctness of the model across all classes. Here, the accuracy is 0.93, which means that the model correctly predicted 93% of all instances in Dry Bean dataset.

The "macro avg" and "weighted avg" are the averages of precision, recall, and F1-score across all classes in Dry Bean dataset. "Macro avg" gives equal weight to each class, while "weighted avg" takes into account the class imbalance by weighting each class's score by its support.

Overall, this classification result suggests that the model has good performance, with high precision, recall, and F1-score for most classes, and a high overall accuracy of 93%.